### PR TITLE
ENH: Robust PCA with improved stochastic gradient descent

### PR DESCRIPTION
**Pull request by tjof2**
A small update to the ORPCA algorithm with a new stochastic gradient descent method that now uses "momentum" to improve the algorithm's robustness with respect to local minima (more [here](http://sebastianruder.com/optimizing-gradient-descent/index.html#momentum)). 

This seems particularly useful at improving the speed of convergence compared to the original `SGD` method, with minimal computational overhead. Almost on a par with the much slower `CF` and `BCD` methods in terms of convergence. This is because it uses memory of the previous samples in a smarter
way.

```
s.decomposition(algorithm='ORPCA',
                output_dimension=3,
                method='MomentumSGD',
                learning_rate=1.1,
                momentum=0.5)
```

### CODE REVIEW

**Comment by tjof2** [comment]
failing tests seem unrelated...

File: hyperspy/learn/rpca.py [hunk 1]
```diff
@@ -221,7 +221,8 @@ def orpca(X, rank, fast=False,
           method=None,
           learning_rate=None,
           init=None,
-          training_samples=None):
+          training_samples=None
```
Comment (thomasaarholt) [inline hyperspy/learn/rpca.py]: comma.

;)

**Comment by tjof2** [comment]
Ok so the Travis test is fixed, AppVeyor is unrelated.

Ready for review.

**Comment by tjof2** [comment]
For reference, AppVeyor error is:

```
%CMD_IN_ENV% move %PYTHON%\sip %PYTHON%\Lib\site-packages\PyQt4\sip
The system cannot find the path specified.
        0 dir(s) moved.
Command exited with code 1
```

**Comment by tjof2** [comment]
@francisco-dlp once this is implemented I intend to update the tutorial notebook (currently PR'ed to the demos repo) with a much clearer discussion of online and robust learning. 

Has the pyqt error in AppVeyor been fixed?

**Comment by francisco-dlp** [comment]
No, the appveyor issue has not been fixed. It is caused by the fact that anaconda now ships with qt5 instead of qt4. Probably simply removing the offending line should fix the issue.

**Comment by tjof2** [comment]
@francisco-dlp  thanks.

**Comment by francisco-dlp** [comment]
The CI problem is now fixed in RELEASE_next_patch, would you like to update your branch?

**Comment by tjof2** [comment]
@francisco-dlp @to266 this has failed on AppVeyor due to something samfire related... :-)

**Comment by francisco-dlp** [comment]
Yes, that's a long standing issue with one of the SAMFire tests that is randomly failing. I'll rerun appveyor.

**Comment by tjof2** [comment]
Thanks @francisco-dlp

**Comment by tjof2** [comment]
Passing AppVeyor at last!

**Comment by francisco-dlp** [comment]
If you merge in RELEASE_next_patch you won't let it down to luck next time you commit something here.

**Comment by thomasaarholt** [comment]
Is this ready to be merged?

**Comment by tjof2** [comment]
@thomasaarholt I've certainly done all I want to do with it!

**Comment by thomasaarholt** [comment]
@francisco-dlp Do you want to merge this?

**Comment by francisco-dlp** [comment]
It needs reviewing and, unfortunately, this week I have no time to review anything other than quick bug fixes.

**Comment by tjof2** [comment]
@francisco-dlp I understand totally!

For anyone who does come to review, a brief summary:

- Expanded documentation on the methods available in ORPCA (e.g. block coordinate descent, stochastic gradient descent, momentum etc.)
- Added a new momentum SGD method which is typically more robust than vanilla SGD as it uses memory from the previous samples, while being faster than the coordinate descent method.

File: hyperspy/learn/rpca.py [hunk 2]
```diff
@@ -238,10 +239,11 @@ def orpca(X, rank, fast=False,
     lambda2 : None | float
         Sparse error regularization parameter.
         If None, set to 1 / sqrt(nsamples)
-    method : None | 'CF' | 'BCD' | 'SGD'
+    method : None | 'CF' | 'BCD' | 'SGD' | 'MomentumSGD'
```
Comment (francisco-dlp) [inline hyperspy/learn/rpca.py]: We follow the [numpy docstring style](https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt) so `{None, 'CF',  'BCD',  'SGD'} instead of `None | 'CF' | 'BCD' | 'SGD'`

File: hyperspy/learn/rpca.py [hunk 3]
```diff
@@ -251,10 +253,13 @@ def orpca(X, rank, fast=False,
         'qr'   - QR-based initialization
         'rand' - Random initialization
         If None, set to 'qr'
-    training_samples : integer
+    training_samples : None | integer
```
Comment (francisco-dlp) [inline hyperspy/learn/rpca.py]: As above.

File: hyperspy/learn/rpca.py [hunk 4]
```diff
@@ -251,10 +253,13 @@ def orpca(X, rank, fast=False,
...
         Specifies the number of training samples to use in
         the 'qr' initialization
         If None, set to 10
+    momentum : None | integer
+        Hyperparameter for 'MomentumSGD' method.
```
Comment (francisco-dlp) [inline hyperspy/learn/rpca.py]: What is a hyperparameter?

**Review by francisco-dlp — CHANGES_REQUESTED** [review]
All fine except for the very minor comments.

File: hyperspy/learn/rpca.py (hunk 4 above)
Comment (tjof2) [reply hyperspy/learn/rpca.py]: I'll clarify in the updated docstring.

**Comment by tjof2** [comment]
@francisco-dlp I think that's everything, caught a few other typos as I went along!
